code/datums/components/drop_shadow.dm

```diff
@@ -0,0 +1,112 @@
+/// Draws a shadow overlay under the attachee
+/datum/component/drop_shadow
+	/// The overlay we are using
+	var/mutable_appearance/shadow
+	/// Extra offset to apply to the shadow
+	var/shadow_offset
+	/// Any temporary extra offsets we are tracking
+	var/additional_offset = 0
+	/// Timer to make sure
+	var/unhide_shadow_timer
+
+/datum/component/drop_shadow/Initialize(icon = 'icons/mob/mob_shadows.dmi', icon_state = SHADOW_MEDIUM, shadow_offset_x = 0, shadow_offset_y = 0)
+	. = ..()
+	if (!ismovable(parent)) // Only being used for mobs at the moment but it seems reasonably likely that we'll want to put it on some effect some time
+		return COMPONENT_INCOMPATIBLE
+
+	src.shadow_offset = shadow_offset_y
+
+	var/atom/movable/movable_parent = parent
+
+	shadow = mutable_appearance(
+		icon,
+		icon_state,
+		layer = BELOW_MOB_LAYER,
+		appearance_flags = KEEP_APART | RESET_TRANSFORM | RESET_COLOR
+	)
+	shadow.pixel_x = shadow_offset_x - movable_parent.pixel_x
+	update_shadow_position()
+
+/datum/component/drop_shadow/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_ATOM_UPDATE_OVERLAYS, PROC_REF(on_update_overlays))
+	RegisterSignals(parent, list(COMSIG_ATOM_FULTON_BEGAN, COMSIG_ATOM_BEGAN_ORBITING), PROC_REF(hide_shadow))
+	RegisterSignals(parent, list(COMSIG_ATOM_FULTON_LANDED, COMSIG_ATOM_STOPPED_ORBITING), PROC_REF(show_shadow))
+	RegisterSignals(parent, list(SIGNAL_ADDTRAIT(TRAIT_SHADOWLESS), SIGNAL_REMOVETRAIT(TRAIT_SHADOWLESS)), PROC_REF(shadowless_trait_updated))
+
+	if (isliving(parent))
+		RegisterSignal(parent, COMSIG_LIVING_POST_UPDATE_TRANSFORM, PROC_REF(on_transform_updated))
+		RegisterSignal(parent, COMSIG_MOB_BUCKLED, PROC_REF(hide_shadow))
+		RegisterSignal(parent, COMSIG_MOB_UNBUCKLED, PROC_REF(show_shadow))
+
+	var/atom/atom_parent = parent
+	atom_parent.update_appearance(UPDATE_OVERLAYS)
+
+/datum/component/drop_shadow/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_ATOM_UPDATE_OVERLAYS,
+		COMSIG_ATOM_FULTON_BEGAN,
+		COMSIG_ATOM_FULTON_LANDED,
+		COMSIG_ATOM_BEGAN_ORBITING,
+		COMSIG_ATOM_STOPPED_ORBITING,
+		COMSIG_LIVING_POST_UPDATE_TRANSFORM,
+		COMSIG_MOB_BUCKLED,
+		COMSIG_MOB_UNBUCKLED,
+		SIGNAL_ADDTRAIT(TRAIT_SHADOWLESS),
+		SIGNAL_REMOVETRAIT(TRAIT_SHADOWLESS),
+	))
+
+/// Repositions the shadow to try and stay under our mob should be at under current conditions
+/datum/component/drop_shadow/proc/update_shadow_position()
+	var/lying_offset = 0
+	if (isliving(parent))
+		var/mob/living/living_parent = parent
+		if (living_parent.rotate_on_lying && living_parent.body_position != STANDING_UP)
+			lying_offset = living_parent.body_position_pixel_y_offset
+		shadow.transform = matrix() * living_parent.current_size
+
+	shadow.pixel_y = -DEPTH_OFFSET - additional_offset - lying_offset + shadow_offset
+	var/atom/atom_parent = parent
+	atom_parent.update_appearance(UPDATE_OVERLAYS)
```

Comment:
should we ignore this call if the trait is not set?

Reply:
Not sure what you mean here.
If you have `TRAIT_SHADOWLESS` we still want the rest of this proc to run so that the shadow is in the right place when it returns.
I could guard calling `update_appearance` with a check for if you have the trait to avoid some proc call overhead (and updating every other overlay for no reason) if you like, although in that case I might as well also do it everywhere else.

Reply:
I was concerned about unneeded overhead yeah. Not because of proc calls but because update_appearance mucks with well, the appearance, and that can be kinda expensive